File: flutter_obdii/lib/core/config_data.dart
// Port of ConfigData.swift — Jim Mittler
// Global configuration data singleton.
// Manages app-wide settings including WiFi connection details,
// auto-connect preferences, connection type, and measurement units.
// Uses SharedPreferences for persistence (replaces @AppStorage).

import 'package:flutter/foundation.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'carplay_bridge.dart';

// ─────────────────────────────────────────────
// MARK: Enums
// ─────────────────────────────────────────────

enum MeasurementUnit {
  metric('Metric'),
  imperial('Imperial');

  const MeasurementUnit(this.displayName);
  final String displayName;

  MeasurementUnit get next =>
      this == MeasurementUnit.metric ? MeasurementUnit.imperial : MeasurementUnit.metric;
}

enum ConnectionType {
  bluetooth('bluetooth'),
  wifi('wifi'),
  demo('demo');

  const ConnectionType(this.rawValue);
  final String rawValue;

  static ConnectionType fromRaw(String raw) =>
      ConnectionType.values.firstWhere((e) => e.rawValue == raw,
          orElse: () => ConnectionType.bluetooth);
}

// ─────────────────────────────────────────────
// MARK: SettingsConfigProviding (protocol)
// ─────────────────────────────────────────────

abstract class SettingsConfigProviding {
  String get wifiHost;
  set wifiHost(String v);
  int get wifiPort;
  set wifiPort(int v);
  bool get autoConnectToOBD;
  set autoConnectToOBD(bool v);
  ConnectionType get connectionType;
  set connectionType(ConnectionType v);
  MeasurementUnit get units;

  void setUnits(MeasurementUnit units);

  // Streams so ViewModels can react
  Stream<MeasurementUnit> get unitsStream;
  Stream<ConnectionType> get connectionTypeStream;
}

// ─────────────────────────────────────────────
// MARK: UnitsProviding (protocol)
// ─────────────────────────────────────────────

abstract class UnitsProviding {
  MeasurementUnit get units;
  Stream<MeasurementUnit> get unitsStream;
}

// ─────────────────────────────────────────────
// MARK: ConfigData singleton
// ─────────────────────────────────────────────

class ConfigData extends ChangeNotifier
    implements SettingsConfigProviding, UnitsProviding {
  static final ConfigData instance = ConfigData._();
  ConfigData._();

  // Persistence keys
  static const _kWifiHost = 'wifiHost';
  static const _kWifiPort = 'wifiPort';
  static const _kAutoConnect = 'autoConnectToOBD';
  static const _kConnectionType = 'connectionType';
  static const _kUnits = 'units';

  // In-memory state
  String _wifiHost = '192.168.0.10';
  int _wifiPort = 35000;
  bool _autoConnectToOBD = true;
  ConnectionType _connectionType = ConnectionType.bluetooth;
  MeasurementUnit _units = MeasurementUnit.metric;

  bool _loaded = false;

  // Broadcast streams (replaces @Published / Combine)
  final _unitsController = ValueNotifier<MeasurementUnit>(MeasurementUnit.metric);
  final _connectionTypeController = ValueNotifier<ConnectionType>(ConnectionType.bluetooth);

  @override
  Stream<MeasurementUnit> get unitsStream =>
      Stream.multi((c) {
        c.add(_unitsController.value);
        void listener() => c.add(_unitsController.value);
        _unitsController.addListener(listener);
        c.onCancel = () => _unitsController.removeListener(listener);
      });

  @override
  Stream<ConnectionType> get connectionTypeStream =>
      Stream.multi((c) {
        c.add(_connectionTypeController.value);
        void listener() => c.add(_connectionTypeController.value);
        _connectionTypeController.addListener(listener);
        c.onCancel = () => _connectionTypeController.removeListener(listener);
      });

  // ── Getters / Setters ────────────────────────

  @override
  String get wifiHost => _wifiHost;
  @override
  set wifiHost(String v) {
    if (v == _wifiHost) return;
    _wifiHost = v;
    _persist(_kWifiHost, v);
    _notifyCarPlaySettingsChanged();
    notifyListeners();
  }

  @override
  int get wifiPort => _wifiPort;
  @override
  set wifiPort(int v) {
    if (v == _wifiPort) return;
    _wifiPort = v;
    _persist(_kWifiPort, v);
    _notifyCarPlaySettingsChanged();
    notifyListeners();
  }

  @override
  bool get autoConnectToOBD => _autoConnectToOBD;
  @override
  set autoConnectToOBD(bool v) {
    if (v == _autoConnectToOBD) return;
    _autoConnectToOBD = v;
    _persist(_kAutoConnect, v);
    _notifyCarPlaySettingsChanged();
    notifyListeners();
  }

  @override
  ConnectionType get connectionType => _connectionType;
  @override
  set connectionType(ConnectionType v) {
    if (v == _connectionType) return;
    _connectionType = v;
    _persist(_kConnectionType, v.rawValue);
    _connectionTypeController.value = v;
    _notifyCarPlaySettingsChanged();
    notifyListeners();
  }

  @override
  MeasurementUnit get units => _units;

  @override
  void setUnits(MeasurementUnit newUnits) {
    if (newUnits == _units) return;
    _units = newUnits;
    _persist(_kUnits, newUnits.name);
    _unitsController.value = newUnits;
    _notifyCarPlaySettingsChanged();
    notifyListeners();
  }

  // ── Load from SharedPreferences ──────────────

  Future<void> load() async {
    if (_loaded) return;
    _loaded = true;
    final prefs = await SharedPreferences.getInstance();

    _wifiHost = prefs.getString(_kWifiHost) ?? '192.168.0.10';
    _wifiPort = prefs.getInt(_kWifiPort) ?? 35000;
    _autoConnectToOBD = prefs.getBool(_kAutoConnect) ?? true;

    final rawConn = prefs.getString(_kConnectionType) ?? 'bluetooth';
    _connectionType = ConnectionType.fromRaw(rawConn);

    final rawUnits = prefs.getString(_kUnits) ?? 'metric';
    _units = MeasurementUnit.values
        .firstWhere((e) => e.name == rawUnits, orElse: () => MeasurementUnit.metric);

    _unitsController.value = _units;
    _connectionTypeController.value = _connectionType;

    notifyListeners();
  }

  // ── Helpers ──────────────────────────────────

  Future<void> _persist(String key, Object value) async {
    final prefs = await SharedPreferences.getInstance();
    if (value is String) await prefs.setString(key, value);
    if (value is int) await prefs.setInt(key, value);
    if (value is bool) await prefs.setBool(key, value);
  }

  void _notifyCarPlaySettingsChanged() {
    CarPlayBridge.settingsChanged(
      units: _units,
      connectionType: _connectionType,
      autoConnectToOBD: _autoConnectToOBD,
      wifiHost: _wifiHost,
      wifiPort: _wifiPort,
    );
  }
}

File: kotlin_obdii/app/src/main/java/com/rheosoft/obdii/core/ConfigData.kt
package com.rheosoft.obdii.core

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/*
enum class MeasurementUnit(val displayName: String) {
    metric("Metric"),
    imperial("Imperial");

    val next: MeasurementUnit
        get() = if (this == metric) imperial else metric
}
*/
enum class ConnectionType(val rawValue: String) {
    bluetooth("bluetooth"),
    wifi("wifi"),
    demo("demo");

    companion object {
        fun fromRaw(raw: String): ConnectionType =
            entries.firstOrNull { it.rawValue == raw } ?: bluetooth
    }
}

enum class GaugesDisplayMode { gauges, list }

interface SettingsConfigProviding {
    var wifiHost: String
    var wifiPort: Int
    var autoConnectToOBD: Boolean
    var connectionType: ConnectionType
    var gaugesDisplayMode: GaugesDisplayMode
    val units: MeasurementUnit
    fun setUnits(units: MeasurementUnit)
    val unitsStream: StateFlow<MeasurementUnit>
    val connectionTypeStream: StateFlow<ConnectionType>
    val gaugesDisplayModeStream: StateFlow<GaugesDisplayMode>
}

interface UnitsProviding {
    val units: MeasurementUnit
    var gaugesDisplayMode: GaugesDisplayMode
    val unitsStream: StateFlow<MeasurementUnit>
    val gaugesDisplayModeStream: StateFlow<GaugesDisplayMode>
}

interface KeyValueStore {
    fun putString(key: String, value: String)
    fun putInt(key: String, value: Int)
    fun putBoolean(key: String, value: Boolean)
    fun getString(key: String): String?
    fun getInt(key: String): Int?
    fun getBoolean(key: String): Boolean?
}

class InMemoryKeyValueStore : KeyValueStore {
    private val map = mutableMapOf<String, Any>()
    override fun putString(key: String, value: String) { map[key] = value }
    override fun putInt(key: String, value: Int) { map[key] = value }
    override fun putBoolean(key: String, value: Boolean) { map[key] = value }
    override fun getString(key: String): String? = map[key] as? String
    override fun getInt(key: String): Int? = map[key] as? Int
    override fun getBoolean(key: String): Boolean? = map[key] as? Boolean
}

object ConfigData : SettingsConfigProviding, UnitsProviding {
    private const val K_WIFI_HOST = "wifiHost"
    private const val K_WIFI_PORT = "wifiPort"
    private const val K_AUTO_CONNECT = "autoConnectToOBD"
    private const val K_CONNECTION_TYPE = "connectionType"
    private const val K_UNITS = "units"
    private const val K_GAUGES_MODE = "gaugesDisplayMode"

    var store: KeyValueStore = InMemoryKeyValueStore()

    override var wifiHost: String = DEFAULT_WIFI_HOST
        set(value) {
            if (field == value) return
            field = value
            store.putString(K_WIFI_HOST, value)
        }

    override var wifiPort: Int = 35000
        set(value) {
            if (field == value) return
            field = value
            store.putInt(K_WIFI_PORT, value)
        }

    override var autoConnectToOBD: Boolean = true
        set(value) {
            if (field == value) return
            field = value
            store.putBoolean(K_AUTO_CONNECT, value)
        }

    override var connectionType: ConnectionType = ConnectionType.bluetooth
        set(value) {
            if (field == value) return
            field = value
            store.putString(K_CONNECTION_TYPE, value.rawValue)
            _connectionTypeFlow.value = value
        }

    override var gaugesDisplayMode: GaugesDisplayMode = GaugesDisplayMode.gauges
        set(value) {
            if (field == value) return
            field = value
            store.putString(K_GAUGES_MODE, value.name)
            _gaugesModeFlow.value = value
        }

    override var units: MeasurementUnit = MeasurementUnit.Metric
        private set

    private val _unitsFlow = MutableStateFlow(MeasurementUnit.Metric)
    private val _connectionTypeFlow = MutableStateFlow(ConnectionType.bluetooth)
    private val _gaugesModeFlow = MutableStateFlow(GaugesDisplayMode.gauges)

    override val unitsStream: StateFlow<MeasurementUnit> = _unitsFlow.asStateFlow()
    override val connectionTypeStream: StateFlow<ConnectionType> = _connectionTypeFlow.asStateFlow()
    override val gaugesDisplayModeStream: StateFlow<GaugesDisplayMode> = _gaugesModeFlow.asStateFlow()

    fun load() {
        wifiHost = store.getString(K_WIFI_HOST) ?: DEFAULT_WIFI_HOST
        wifiPort = store.getInt(K_WIFI_PORT) ?: 35000
        autoConnectToOBD = store.getBoolean(K_AUTO_CONNECT) ?: true
        connectionType = ConnectionType.fromRaw(store.getString(K_CONNECTION_TYPE) ?: "bluetooth")
        units = MeasurementUnit.entries.firstOrNull { it.name == (store.getString(K_UNITS) ?: "metric") }
            ?: MeasurementUnit.Metric
        gaugesDisplayMode = GaugesDisplayMode.entries.firstOrNull { it.name == (store.getString(K_GAUGES_MODE) ?: "gauges") }
            ?: GaugesDisplayMode.gauges
        _unitsFlow.value = units
        _connectionTypeFlow.value = connectionType
        _gaugesModeFlow.value = gaugesDisplayMode
    }

    override fun setUnits(units: MeasurementUnit) {
        if (this.units == units) return
        this.units = units
        store.putString(K_UNITS, units.name)
        _unitsFlow.value = units
    }

    fun resetForTests() {
        wifiHost = DEFAULT_WIFI_HOST
        wifiPort = 35000
        autoConnectToOBD = true
        connectionType = ConnectionType.bluetooth
        setUnits(MeasurementUnit.Metric)
    }
}
